Check for use of `InstId`s from the wrong `SemIR::File`

dwblaikie opened the pull request:
Use the `CheckIRId` as a unique identifier for the scope of an `InstId` - if an `InstId` is created within the scope of one `CheckIRId` it must not be used in the scope of a different `CheckIRId`.

This is achieved without extra storage, but with false negatives for large inputs.

When an `InstId` is created, the original index of the `Inst` is XORed with a tag derived from the `CheckIRId` to produce the final `InstId`. When the `InstId` is used, the expected tag is XORed with the `InstId` to get back to the original index - if the tags don't match, the resulting index will be corrupted, likely too large - resulting in an out of bounds index CHECK-failure.

(the tag value is derived as such:
* take the CheckIRId
* left shift one bit (padding zero)
* left shift another bit (padding 1 - used to signify that the resulting `InstId` has a tag combined into it)
* reverse the bits

In this way, the tag is unlikely to overlap with the index for small test cases - making it possible to separate out the `CheckIRId` from the index in these cases to provide more meaningful debugging/CHECK messages, and more informative `SemIR` textual dumping that can now include the `CheckIRId` along with the `Inst`'s index in the name of an `inst`)

The test churn here is improved printing as tagged `InstId`s can now, with best effort (more likely for small test cases where the `CheckIRId` and the `Inst` index aren't at risk of overlapping from the high and low bits), render the `CheckIRId` as part of the inst's name. Going from `instNN` to `irMM.instNN`.

dwblaikie committed:
WIP

dwblaikie committed:
Add InstStore to Constants so the id can be validated

Otherwise invalid InstIds passed to canstants would cause
arbitrary/large memory allocation as the constant storage would be grown
to fit the giant invalid index.

dwblaikie committed:
Fix

dwblaikie committed:
Best effort attempt to decompose an id into unit+index in the absence of any context

useful for dumping given only an id without the container or SemIR::File
it might've come from

dwblaikie committed:
Refactor best-effort decomposition for reuse with CARBON_DCHECK

The more-informative check failure (reproducing the original bug that
motivated this extra checking) looks like:

CHECK failure at ./toolchain/base/value_store.h:295: index < size_:
  Untagged index was outside of container range.
  Possibly tagged index 2113929262.
  Best-effort decomposition: CheckIRId: 30, index: 46.
  The CheckIRIdTag for this container is: 29

Not sure how this should be phrased - it all feels a bit awkward.

Maybe something like "CheckIRId was 30, should be 29. Index is 46"?
Maybe print the tagged index as hex?

dwblaikie committed:
Minor updates/fixes from rebase

dwblaikie committed:
Non-automatic test updates

zygoloid commented on toolchain/base/value_store.h:
```diff
@@ -314,7 +415,7 @@ class ValueStore
   // Converts an id into an index into the set of chunks, and an offset into
   // that specific chunk. Looks for index overflow in non-optimized builds.
-  static auto IdToChunkIndices(IdType id) -> std::pair<int32_t, int32_t> {
+  static auto IdToChunkIndices(int32_t index) -> std::pair<int32_t, int32_t> {
```
Maybe rename to `RawIndexToChunkIndices`? Might also be worth keeping an `IdToChunkIndices` overload that takes an `IdType` since many of the callers want that.

dwblaikie replied:
Ah, yep - 728a04b53a4c9cfee3849f1e8f873b95da3d417e 
Though `IdToChunkIndices` becomes a non-static member, since it needs to access the tag to unswizzle things. But that's fine/works for existing callers.

zygoloid commented on toolchain/base/value_store.h:
```diff
@@ -32,6 +32,87 @@ class ValueStoreNotPrintable {};
 
 }  // namespace Internal
 
+struct CheckIRIdTag {
```
I'd be inclined to rename this to something more general, perhaps just `IdTag`. For the `ValueStore`s that have nothing to do with `CheckIR`s, having a `CheckIRIdTag` parameter is a bit strange, but a general `IdTag` seems reasonable and could in principle be useful (we could use something else as the tag in places where that makes sense).

While I like `Tag` and for me the connection to pointer tagging works well, if you find it insufficiently descriptive another possible name would be something like `IdSpace`.

dwblaikie replied:
`IdTag` works for me, 8ae0288c9be86b32b6a3c879951ef5f04fcded11

zygoloid commented on toolchain/sem_ir/constant.h:
```diff
@@ -122,18 +125,17 @@ class ConstantValueStore {
   // Sets the constant value of the given instruction, or sets that it is known
   // to not be a constant.
   auto Set(InstId inst_id, ConstantId const_id) -> void {
-    CARBON_DCHECK(inst_id.index >= 0);
-    if (static_cast<size_t>(inst_id.index) >= values_.size()) {
-      values_.Resize(inst_id.index + 1, default_);
+    auto index = insts_ ? insts_->GetRawIndex(inst_id) : inst_id.index;
+    if (static_cast<size_t>(index) >= values_.size()) {
+      values_.Resize(index + 1, default_);
     }
     values_.Get(inst_id) = const_id;
   }
```
It looks like if `insts_` isn't set, we could easily allocate a 4-8GB value store here if `inst_id` is tagged.

It would be good to understand how we build a `ConstantValueStore` with no `InstStore` -- this is presumably happening in the call from `import_ref.cpp`, and I guess we're calling that function before we build the `SemIR` for the dependency. Maybe it's possible to determine and pass in the `CheckIRId` from there.

(As far as I'm aware, there are only two places where we create `ConstantValueStore`s -- we have one for a file, and then each file has one for each `ImportIRId` to determine the constant values of that other file's instructions in this file's constant ID space. Both kinds should have tagged `InstId`s.)

dwblaikie replied:
> It looks like if `insts_` isn't set, we could easily allocate a 4-8GB value store here if `inst_id` is tagged.

Yep... had some of those failures while debugging this work in progress.

> It would be good to understand how we build a ConstantValueStore with no InstStore -- this is presumably happening in the call from import_ref.cpp, and I guess we're calling that function before we build the SemIR for the dependency. Maybe it's possible to determine and pass in the CheckIRId from there.

Yep, that's the one that I think I came across.

> (As far as I'm aware, there are only two places where we create ConstantValueStores -- we have one for a file, and then each file has one for each ImportIRId to determine the constant values of that other file's instructions in this file's constant ID space. Both kinds should have tagged InstIds.)

Hm, OK.

Looks like the case where the `ConstantValueStore`s in `import_ref.cpp` are created without a `SemIR::File` are for the two special `ImportIR`s: `ImportIRId::ApiForImpl` and `ImportIRId::Cpp` - in both those cases, it seems the `ImportIR` never gets a `SemIR::File`?

Hmm, "A null entry is added if there is none, in which case this ID should not show up in instructions."

I see, so empty placeholder ones are added, and then optionally replaced if needed. It's the empty ones that are created without a `SemIR::File`?

Hmm - maybe the `ApiForImpl` one has no `SemIR::File` if it's unused. But for `Cpp` -  it looks like it never has a `SemIR::File`? Am I understanding that correctly?

If so, what'd you reckon we do for the `ConstantValueStore`s associated with these two (well, associated with `Cpp` in any case, and associated with `ApiForImpl` when there is no api for impl - I guess in this latter case some kind of nonce `IdTag` that's never valid/check-fails immediately on use would suffice, maybe the same is true for `Cpp`, that nothing should actually get stored in this `ConstantValueStore`?)

Yeah, looks like in these cases the ConstantValueStore is never used - so I can change the conditional use of `insts_` to a `CARBON_CHECK` that `insts_` is non-null and then use it unconditionally. Done in 280ed8ce4e28f2fb570670277a12e94418670e62

But I'm open to better ideas here... maybe some way not to have these unused `ConstantValueStore`s at all? Or make them even more explicitly/directly explosive-if-used.

zygoloid replied:
I see, so the problem is specifically the calls to `InternalAddImportIR` from `SetSpecialImportIR`, which are just creating placeholders and not adding a real value store at all.

Perhaps we could add another `ConstantValueStore` constructor to build an invalid value store, and use that instead of calling `InternalAddImportIR` from `SetSpecialImportIR` in the case where there is no corresponding `SemIR`.

dwblaikie replied:
Gave that a go in  dd1911f

zygoloid commented on toolchain/base/fixed_size_value_store.h:
```diff
@@ -73,7 +76,8 @@ class FixedSizeValueStore {
   explicit FixedSizeValueStore(const ValueStoreT& size_source,
-                               ConstRefType default_value) {
+                               ConstRefType default_value)
+      : tag_(GetCheckIRIdTag(size_source)) {
```
Do you know what types get used as `ValueStoreT` here? I wonder if we can use `size_source.GetCheckIRIdTag()` here instead of going via the helper function.

dwblaikie replied:
Currently, at least - `ValueStoreT` can also be
* `SpecificStore` here https://github.com/carbon-language/carbon-lang/blob/cac3578b86f0ff08247e07b08e513cda9678f535/toolchain/lower/specific_coalescer.cpp#L15-L17
* `GenericStore` here https://github.com/carbon-language/carbon-lang/blob/cac3578b86f0ff08247e07b08e513cda9678f535/toolchain/lower/file_context.cpp#L57

In theory if we push this id-validation into other containers beyond insts (which I'd be happy to - if you/folks are on board with it) - then eventually we'd have it for specific & generic store, and they'd all have the `GetIdTag` (nee `GetCheckIRIdTag`) function and there'd be no need for this template specialization indirection.

Certainly open to ideas.

zygoloid replied:
For now, could we add a `GetIdTag` member function to those two stores and make it return `IdTag()` to explicitly indicate that no tagging is happening? Then I think we could use `size_source.GetIdTag()` here -- I think that'd be simpler overall than the function template specialization.

dwblaikie replied:
Ah, sure - yeah, the ADL thing was overly cautious about touching other things - that I'm going to touch anyway most likely.

Done in 306fcc6

zygoloid commented on toolchain/sem_ir/ids.cpp:
```diff
@@ -16,7 +16,13 @@ auto InstId::Print(llvm::raw_ostream& out) const -> void {
+    if (check_ir_id == -1) {
+      IdBase::Print(out);
+    } else {
+      out << "inst." << check_ir_id << '.' << simple_index;
```
Perhaps we could print this as `ir3.inst5` -- that seems like it'd be easier to regex match and maybe also a little clearer about what the two numbers are and how this relates to the format without an ID.

dwblaikie replied:
I like it - 6649f474ce60433b14a56b9cfe9bbd25fe0a0a65

dwblaikie committed:
Rename `IdToChucknIndices` to `RawIndexToChunkIndices` to be more accurate

Add back in the original `IdToChunkIndices` but as a non-static member
(since it's needed to un-swizzle the CheckIRIdTag back out of the Id to
get to the raw index)

dwblaikie committed:
Rename `CheckIRIdTag` to `IdTag`

dwblaikie committed:
CARBON_CHECK that used ConstantValueStores have an associated InstStore

dwblaikie committed:
Change best-effort inst name printing to `irN.instM`

dwblaikie commented:
Ping @zygoloid - further thoughts?

zygoloid commented on toolchain/sem_ir/constant.h:
```diff
@@ -109,8 +109,11 @@ struct SymbolicConstant : Printable<SymbolicConstant> {
+  explicit ConstantValueStore(ConstantId default_value, const InstStore* insts)
+      : default_(default_value),
+        values_(insts ? insts->GetIdTag()
+                      : IdTag(0, std::numeric_limits<int32_t>::max())),
```
```suggestion
                      : IdTag(),
```

zygoloid reviewed (commented):
Thanks, a couple of comments but basically LGTM.

(Looks like it's worth making some updates to the first comment on the PR before that becomes the commit message.)

I forget -- have you done any performance testing here? I'm not really expecting much difference but value stores are pretty hot so it seems worth checking.

dwblaikie committed:
Simplify IdTag default construction

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

dwblaikie committed:
Add missing ')' and format

dwblaikie committed:
Put stub `GetIdTag` in SpecificStore, removing ADL-based `GetIdTag`

dwblaikie committed:
Add a ConstantValueStore ctor that explicitly makes the store unusable

dwblaikie reviewed (commented):
> Thanks, a couple of comments but basically LGTM.
> 
> (Looks like it's worth making some updates to the first comment on the PR before that becomes the commit message.)

Ah, yep - updated.

> I forget -- have you done any performance testing here? I'm not really expecting much difference but value stores are pretty hot so it seems worth checking.

I have not - any recommended benchmarks/mechanisms for performance testing changes?

dwblaikie commented:
yeah, got some performance regressions I'll need to look into :/
```
 Benchmark                                        ┃          CPU Time           ┃          Bytes          ┃          Lines          ┃         Tokens
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━
 BM_CompileApiFileDenseDecls<Phase::Lex>/256..... │      ??          p=0.291    │ ~                       │ ~                       │ ~
                                        baseline: │    206.5   µs  ±   1.400%   │    25.51  M ±   1.420%  │   944.5   k ±   1.420%  │     5.333 M ±   1.420%
                                      experiment: │    207     µs  ±   1.032%   │    25.44  M ±   1.022%  │   942     k ±   1.022%  │     5.319 M ±   1.022%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Lex>/1024.... │      ??          p=0.26     │ ~                       │ ~                       │ ~
                                        baseline: │    315.5   µs  ±   2.879%   │   103     M ±   2.798%  │     3.103 M ±   2.798%  │    18.23  M ±   2.798%
                                      experiment: │    319.7   µs  ±   0.702%   │   101.6   M ±   0.707%  │     3.062 M ±   0.707%  │    17.99  M ±   0.707%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Lex>/4096.... │ 👎   1.086%      p=0.0137   │ ~                       │ ~                       │ ~
                                        baseline: │    751.3   µs  ±   1.788%   │   188.9   M ±   1.820%  │     5.361 M ±   1.820%  │    31.76  M ±   1.820%
                                      experiment: │    759.5   µs  ±   2.398%   │   186.8   M ±   2.341%  │     5.304 M ±   2.341%  │    31.42  M ±   2.341%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Lex>/16384... │      ??          p=0.0782   │ ~                       │ ~                       │ ~
                                        baseline: │      2.584 ms  ±   4.567%   │   236.4   M ±   4.370%  │     6.312 M ±   4.370%  │    37.45  M ±   4.370%
                                      experiment: │      2.656 ms  ±   2.913%   │   230     M ±   2.831%  │     6.141 M ±   2.831%  │    36.44  M ±   2.831%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Lex>/65536... │      ??          p=0.181    │ ~                       │ ~                       │ ~
                                        baseline: │     10.4   ms  ±   6.726%   │   242.7   M ±   6.305%  │     6.294 M ±   6.305%  │    37.37  M ±   6.305%
                                      experiment: │     10.99  ms  ±   5.948%   │   229.7   M ±   6.324%  │     5.957 M ±   6.324%  │    35.37  M ±   6.324%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Lex>/262144.. │ 👎   4.676%      p=0.0166   │ ~                       │ ~                       │ ~
                                        baseline: │     46.85  ms  ±  10.449%   │   218.2   M ±   9.467%  │     5.596 M ±   9.467%  │    33.22  M ±   9.467%
                                      experiment: │     49.04  ms  ±  10.336%   │   208.4   M ±   9.368%  │     5.345 M ±   9.368%  │    31.74  M ±   9.368%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Parse>/256... │      ??          p=0.438    │ ~                       │ ~                       │ ~
                                        baseline: │    238.4   µs  ±   5.045%   │    22.09  M ±   4.803%  │   817.9   k ±   4.803%  │     4.618 M ±   4.803%
                                      experiment: │    240     µs  ±   1.518%   │    21.94  M ±   1.495%  │   812.4   k ±   1.495%  │     4.587 M ±   1.495%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Parse>/1024.. │      ??          p=0.0572   │ ~                       │ ~                       │ ~
                                        baseline: │    461.1   µs  ±   1.573%   │    70.46  M ±   1.549%  │     2.123 M ±   1.549%  │    12.47  M ±   1.549%
                                      experiment: │    464.8   µs  ±   0.727%   │    69.89  M ±   0.732%  │     2.106 M ±   0.732%  │    12.37  M ±   0.732%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Parse>/4096.. │ 👎   1.825%      p=0.000249 │ ~                       │ ~                       │ ~
                                        baseline: │      1.302 ms  ±   0.605%   │   109     M ±   0.601%  │     3.093 M ±   0.601%  │    18.32  M ±   0.601%
                                      experiment: │      1.326 ms  ±   1.662%   │   107     M ±   1.635%  │     3.038 M ±   1.635%  │    18     M ±   1.635%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Parse>/16384. │ 👎   1.712%      p=0.0166   │ ~                       │ ~                       │ ~
                                        baseline: │      4.897 ms  ±   2.576%   │   124.8   M ±   2.644%  │     3.331 M ±   2.644%  │    19.76  M ±   2.644%
                                      experiment: │      4.981 ms  ±   2.989%   │   122.7   M ±   2.902%  │     3.275 M ±   2.902%  │    19.43  M ±   2.902%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Parse>/65536. │      ??          p=0.398    │ ~                       │ ~                       │ ~
                                        baseline: │     19.87  ms  ±   3.185%   │   127.1   M ±   3.087%  │     3.295 M ±   3.087%  │    19.56  M ±   3.087%
                                      experiment: │     20.36  ms  ±   4.434%   │   124     M ±   4.640%  │     3.216 M ±   4.640%  │    19.09  M ±   4.640%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Parse>/262144 │ 👎   5.161%      p=0.00388  │ ~                       │ ~                       │ ~
                                        baseline: │     83.73  ms  ±   2.650%   │   122.1   M ±   2.599%  │     3.131 M ±   2.599%  │    18.59  M ±   2.599%
                                      experiment: │     88.05  ms  ±   4.060%   │   116.1   M ±   4.232%  │     2.977 M ±   4.232%  │    17.67  M ±   4.232%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Check>/256... │ 👎   9.785%      p=0.000249 │ ~                       │ ~                       │ ~
                                        baseline: │     17.92  ms  ±   0.850%   │   293.9   k ±   0.857%  │    10.88  k ±   0.857%  │    61.43  k ±   0.857%
                                      experiment: │     19.68  ms  ±   0.994%   │   267.7   k ±   1.003%  │     9.911 k ±   1.003%  │    55.96  k ±   1.003%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Check>/1024.. │ 👎   9.446%      p=0.000557 │ ~                       │ ~                       │ ~
                                        baseline: │     19.28  ms  ±   0.896%   │     1.685 M ±   0.888%  │    50.77  k ±   0.888%  │   298.3   k ±   0.888%
                                      experiment: │     21.1   ms  ±   0.452%   │     1.539 M ±   0.450%  │    46.39  k ±   0.450%  │   272.6   k ±   0.450%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Check>/4096.. │ 👎   7.604%      p=0.000428 │ ~                       │ ~                       │ ~
                                        baseline: │     24.8   ms  ±   0.958%   │     5.722 M ±   0.967%  │   162.4   k ±   0.967%  │   962.2   k ±   0.967%
                                      experiment: │     26.68  ms  ±   0.828%   │     5.317 M ±   0.821%  │   151     k ±   0.821%  │   894.2   k ±   0.821%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Check>/16384. │ 👎   5.282%      p=0.000328 │ ~                       │ ~                       │ ~
                                        baseline: │     47.99  ms  ±   1.808%   │    12.73  M ±   1.776%  │   339.9   k ±   1.776%  │     2.017 M ±   1.776%
                                      experiment: │     50.53  ms  ±   2.543%   │    12.09  M ±   2.480%  │   322.8   k ±   2.480%  │     1.916 M ±   2.480%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Check>/65536. │ 👎   3.614%      p=0.00153  │ ~                       │ ~                       │ ~
                                        baseline: │    146.8   ms  ±   1.072%   │    17.2   M ±   1.061%  │   446     k ±   1.061%  │     2.648 M ±   1.061%
                                      experiment: │    152.1   ms  ±   1.689%   │    16.6   M ±   1.661%  │   430.5   k ±   1.661%  │     2.556 M ±   1.661%
                                                  │                             │                         │                         │
 BM_CompileApiFileDenseDecls<Phase::Check>/262144 │ 👎   3.955%      p=0.00743  │ ~                       │ ~                       │ ~
                                        baseline: │    586.6   ms  ±   1.675%   │    17.42  M ±   1.647%  │   446.9   k ±   1.647%  │     2.653 M ±   1.647%
                                      experiment: │    609.8   ms  ±   1.902%   │    16.76  M ±   1.867%  │   429.9   k ±   1.867%  │     2.552 M ±   1.867%
                                                  │                             │                         │                         │                                                  
```

chandlerc commented:
FWIW, those are pretty small regressions, and there were up to 5% swings in the parse benchmarks so might be some noise in them too. Can look for anything obvious, but may not be worth stressing too much about this. IT's not like a 25% or 50% regression.

dwblaikie committed:
Merge branch 'trunk' into debug_checks_checkirid_matching_value_store

dwblaikie commented:
> FWIW, those are pretty small regressions, and there were up to 5% swings in the parse benchmarks so might be some noise in them too. Can look for anything obvious, but may not be worth stressing too much about this. IT's not like a 25% or 50% regression.

Fair - yeah, better understood the perf report and how for higher numbers of iterations the difference went down, and the noise. Ran some perf diffs and didn't find anything really standout. So going ahead with this as-is for now.
